core/object/message_queue.h

```diff
@@ -33,26 +33,45 @@
 
 #include "core/object/object_id.h"
 #include "core/os/thread_safe.h"
+#include "core/templates/local_vector.h"
+#include "core/templates/paged_allocator.h"
 #include "core/variant/variant.h"
 
 class Object;
 
-class MessageQueue {
-	_THREAD_SAFE_CLASS_
-
+class CallQueue {
+public:
 	enum {
-		DEFAULT_QUEUE_SIZE_KB = 4096
+		PAGE_SIZE_BYTES = 4096
 	};
 
+private:
 	enum {
 		TYPE_CALL,
 		TYPE_NOTIFICATION,
 		TYPE_SET,
+		TYPE_END, // End marker.
 		FLAG_SHOW_ERROR = 1 << 14,
-		FLAG_MASK = FLAG_SHOW_ERROR - 1
+		FLAG_MASK = FLAG_SHOW_ERROR - 1,
+	};
 
+	struct Page {
+		uint8_t data[PAGE_SIZE_BYTES];
 	};
 
+	Mutex mutex;
+	typedef PagedAllocator<Page, true> Allocator;
+
+	Allocator *allocator = nullptr;
+	bool allocator_is_custom = false;
+
+	LocalVector<Page *> pages;
+	LocalVector<uint32_t> page_messages;
```

Comment:
I believe that some arithmetic involving `pages_used`, page size and `page_offset` could replace the use cases of this array.

Reply:
It could, and one alternative is to rework the message format and use a terminator case (as an extra value of the message type enum) but it would make the code more complex, so I think this is just simpler.

Reply:
Not the end of the world. I was just thinking about cache pollution, etc., but you never know until you measure in practice.